cpp/include/raft/util/bitset.cuh

```diff
@@ -0,0 +1,202 @@
+
+  /**
+   * @brief Construct a new bitset object with a list of indices to unset.
+   *
+   * @param res RAFT resources
+   * @param mask_index List of indices to unset in the bitset
+   * @param bitset_len Length of the bitset
+   */
+  bitset(const raft::resources& res,
+         raft::device_vector_view<const IdxT, IdxT> mask_index,
+         IdxT bitset_len)
+    : bitset_{raft::make_device_vector<BitsetT, IdxT>(
+        res, raft::ceildiv(bitset_len, bitset_element_size))}
+  {
+    RAFT_EXPECTS(mask_index.extent(0) <= bitset_len, "Mask index cannot be larger than bitset len");
+    cudaMemsetAsync(bitset_.data_handle(),
+                    0xff,
+                    bitset_.size() * sizeof(BitsetT),
+                    resource::get_cuda_stream(res));
+    bitset_unset(res, view(), mask_index);
+  }
+
+  /**
+   * @brief Construct a new bitset object
+   *
+   * @param res RAFT resources
+   * @param bitset_len Length of the bitset
+   */
+  bitset(const raft::resources& res, IdxT bitset_len)
+    : bitset_{raft::make_device_vector<BitsetT, IdxT>(
+        res, raft::ceildiv(bitset_len, bitset_element_size))}
+  {
+    cudaMemsetAsync(bitset_.data_handle(),
+                    0xff,
+                    bitset_.size() * sizeof(BitsetT),
+                    resource::get_cuda_stream(res));
+  }
+  // Disable copy constructor
+  bitset(const bitset&)            = delete;
+  bitset(bitset&&)                 = default;
+  bitset& operator=(const bitset&) = delete;
+  bitset& operator=(bitset&&)      = default;
+
+  /**
+   * @brief Create a device-usable view of the bitset.
+   *
+   * @return bitset_view<IdxT>
+   */
+  inline auto view() -> raft::utils::bitset_view<IdxT> { return bitset_view<IdxT>(bitset_.view()); }
+  [[nodiscard]] inline auto view() const -> raft::utils::bitset_view<IdxT>
+  {
+    return bitset_view<IdxT>(bitset_.view());
+  }
+
+ private:
+  raft::device_vector<BitsetT, IdxT> bitset_;
+};
+
+/**
+ * @brief Function to unset a list of indices in a bitset.
+ *
+ * @tparam IdxT Indexing type used. Default is uint32_t.
+ * @param res RAFT resources
+ * @param bitset_view_ View of the bitset
+ * @param mask_index indices to remove from the bitset
+ */
+template <typename IdxT>
+void bitset_unset(const raft::resources& res,
+                  raft::utils::bitset_view<IdxT> bitset_view_,
+                  raft::device_vector_view<const IdxT, IdxT> mask_index)
+{
+  auto* bitset_ptr = bitset_view_.get_bitset_ptr();
+  thrust::for_each_n(resource::get_thrust_policy(res),
+                     mask_index.data_handle(),
+                     mask_index.extent(0),
+                     [bitset_ptr] __device__(const IdxT sample_index) {
+                       const IdxT bit_element = sample_index / 32;
+                       const IdxT bit_index   = sample_index % 32;
+                       const uint32_t bitmask = ~(1 << bit_index);
+                       atomicAnd(bitset_ptr + bit_element, bitmask);
```

Comment:
Can we avoid the atomic by constructing each `bit_element` that we need to touch and then doing a bitwise AND across a whole element (or even several adjacent elements)?

Reply:
I initially implemented a custom kernel to create the bitset element in shared memory then doing one write to global memory but it was slower than the current approach.
I added benchmarks so that we can iterate later on potential performance improvements.